# Polynomial Regression & Pipeline



## Mục tiêu

Sau khi hoàn thành lab này, bạn có thể:

- Phát triển mô hình hồi quy đa thức sử dụng pipeline
- Hiểu một số thước đo số để xác định mức độ phù hợp của mô hình


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
%matplotlib inline

Trong lab này, học viên có thể sử dụng dữ liệu đã được làm sạch từ lab 6&7 để phục vụ cho việc phát triển mô hình:

In [2]:
# path of data 
path = 'L10_datasets/clean_df.csv'
df = pd.read_csv(path)
df.head()

,symboling,normalized-losses,make,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,length,width,...,peak-rpm,city-mpg,highway-L/100km,price,city-L/100km,horsepower-binned,fuel-type-diesel,fuel-type-gas,aspiration-std,aspiration-turbo
0,3,122,alfa-romero,two,convertible,rwd,front,88.6,0.811148,0.890278,...,5000.0,21,8.703704,13495.0,11.190476,Low,0,1,True,False
1,3,122,alfa-romero,two,convertible,rwd,front,88.6,0.811148,0.890278,...,5000.0,21,8.703704,16500.0,11.190476,Low,0,1,True,False
2,1,122,alfa-romero,two,hatchback,rwd,front,94.5,0.822681,0.909722,...,5000.0,19,9.038462,16500.0,12.368421,Medium,0,1,True,False
3,2,164,audi,four,sedan,fwd,front,99.8,0.848630,0.919444,...,5500.0,24,7.833333,13950.0,9.791667,Low,0,1,True,False
4,2,164,audi,four,sedan,4wd,front,99.4,0.848630,0.922222,...,5500.0,18,10.681818,17450.0,13.055556,Low,0,1,True,False


<h2>Phần 1: Polynomial Regression and Pipeline</h2>


<b>Hôi quy đa thức là gì?</b> Là một trường hợp đặc biệt trong các bài toán hồi quy:
Được sử dụng để mô tả các điểm phân bố, phân tán dưới dạng đường cong

Sử dụng mối quan hệ giữa các biến độc lập x và biến phụ thuộc y được biểu diễn dưới dạng đa thức bậc n và tìm cách tốt nhất vẽ một đường qua các điểm dữ liệu sao cho tối ưu và phù hợp nhất.

Sau đây là một số ví dụ về hồi quy đa thức.

<center><b>Quadratic - Đa thức bậc 2</b></center>

$$
\hat{Y} = a + b_1 X + b_2 X^2
$$

<center><b>Cubic - Đa thức bậc 3</b></center>

$$
\hat{Y} = a + b_1 X + b_2 X^2 + b_3 X^3
$$

<center><b>Đa thức bậc cao hơn:</b></center>

$$
Y = a + b_1 X + b_2 X^2 + b_3 X^3 + \dots
$$


Biểu thức cho hàm đa thức đa biến trở nên phức tạp hơn rất nhiều so với hồi quy đa thức một biến thông thường. Ví dụ: biểu thức của đa thức bậc hai, có hai biến được cho bởi:

$$
Yhat = a + b_1 X_1 +b_2 X_2 +b_3 X_1 X_2+b_4 X_1^2+b_5 X_2^2
$$


Chúng ta đã thấy trước đó rằng mô hình tuyến tính đa biến kết quả khá tốt khi sử dụng bốn đặc điểm `horsepower`, `curb-weight`, `engine-size`, `highway-L/100km` làm biến dự đoán. Hãy xem liệu chúng ta có thể thử điều chỉnh mô hình đa thức cho dữ liệu hay không:

In [3]:
Z = df[['horsepower', 'curb-weight', 'engine-size', 'highway-L/100km']]

In [4]:
Y = df[['price']]

Chúng ta có thể thực hiện một phép biến đổi đa thức trên nhiều biến đầu vào. Đầu tiên, chúng ta cần phải import thư viện liên quan

In [5]:
from sklearn.preprocessing import PolynomialFeatures

Chúng ta tạo một mô hình Polynom2

In [6]:
pr=PolynomialFeatures(degree=2)
pr

PolynomialFeatures()

In [7]:
Z_pr=pr.fit_transform(Z)

Kích thước ban đầu của bộ dữ liệu bao gồm 4 features và 201 samples dữ liệu

In [8]:
Z.shape

(201, 4)

Sau khi biến đổi, của bộ dữ liệu mới bao gồm 15 features và 201 samples dữ liệu

In [9]:
Z_pr.shape

(201, 15)

### Bài tập 1a

<b><b>Tạo một linear regression object, sau đó huấn luyện mô hình sử dụng bộ data đã được biến đổi</b></b>

In [10]:
# Code của bạn ở đây

### Bài tập 1b

<b><b>Vẽ biểu đồ phân bố để đánh giá mô hình</b></b>

In [11]:
# Code của bạn ở đây

Chúng ta có thể thấy, mô hình đa thức phù hợp với dữ liệu tốt hơn.

<h2>Pipeline</h2>


<p>Data Pipelines đơn giản hóa các bước xử lý dữ liệu. Trong phần này chúng ta sử dụng module <b>Pipeline</b> để tạo Pipeline. Chúng ta cũng sử dụng <b>StandardScaler</b> làm một bước trong quy trình của mình.</p>

In [12]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

Chúng ta tạo pipeline bằng cách tạo danh sách các tuples dữ liệu bao gồm tên của mô hình hoặc công cụ ước tính và hàm tương ứng của nó.

In [13]:
Input=[('scale',StandardScaler()), 
       ('polynomial', PolynomialFeatures(include_bias=False)), 
       ('model',LinearRegression())]

Đưa danh sách các bước vừa tạo ở trên làm input đầu vào cho `Pipeline`

In [14]:
pipe=Pipeline(Input)
pipe

Pipeline(steps=[('scale', StandardScaler()),
                ('polynomial', PolynomialFeatures(include_bias=False)),
                ('model', LinearRegression())])

Chúng ta có thể chuẩn hóa dữ liệu, thực hiện chuyển đổi và điều chỉnh mô hình cùng một lúc

In [15]:
pipe.fit(Z,Y)

Pipeline(steps=[('scale', StandardScaler()),
                ('polynomial', PolynomialFeatures(include_bias=False)),
                ('model', LinearRegression())])

Tương tự, chúng ta có thể chuẩn hóa dữ liệu, thực hiện chuyển đổi và đưa ra dự đoán đồng thời

In [16]:
ypipe=pipe.predict(Z)
ypipe[0:5]

array([[12395.58706934],
       [12395.58706934],
       [18422.45365339],
       [ 9979.7757077 ],
       [17551.90086636]])

### Bài tập 2a

<b><b>Vẽ biểu đồ phân bố để đánh giá mô hình sử dụng pipeline phía trên</b></b>

In [17]:
# Code của bạn ở đây

### Bài tập 2b

<b><b>Tạo một pipeline bao gồm các bước: chuẩn hóa dữ liệu bằng MinMaxScaler, thực hiện chuyển đổi và sau đó thực hiện dự đoán bằng mô hình hồi quy tuyến tính và sau đó vẽ biểu đồ phân bố để đánh giá mô hình sử dụng pipeline phía trên.</b></b>

In [18]:
# Code của bạn ở đây

In [19]:
# Code của bạn ở đây

<h2>Phần 2: Đánh giá mức độ hiệu quả của mô hình</h2>


<p>Khi đánh giá các mô hình của mình, chúng ta không chỉ muốn trực quan hóa kết quả mà còn muốn có một thước đo định lượng để xác định mức độ chính xác của mô hình.</p>

<p>2 chỉ số đo lường quan trọng để xác định sự phù hợp của một mô hình:</p>
<ul>
    <li><b>Mean Squared Error (MSE)</b></li>
    <li><b>R^2 / R-squared</b></li>
</ul>
    

<b>Mean Squared Error (MSE)</b>

<p>Sai số bình phương trung bình đo mức trung bình của các bình phương sai số, nghĩa là chênh lệch giữa giá trị thực tế (y) và giá trị ước tính (ŷ).</p>

<b>R-squared</b>
<p>R-squared cho ta biết mức độ các biến đầu vào sẽ giải thích được bao nhiêu phần trăm các biến mục tiêu</p>
    
<p>R-squared càng lớn thì mô hình càng tốt. Khi R-squared bằng 95% điều đó có nghĩa rằng các biến đầu vào đã giải thích được 95% sự biến động của biến mục tiêu</p>

<h3>Mô hình 1: Mô hình Linear Regression</h3>


In [20]:
X = df[['horsepower']]

Tính toán chỉ số R^2


In [21]:
# horsepower_fit
lm = LinearRegression()
lm.fit(X, Y)
# Find the R^2
print('Chỉ số R-square = ', lm.score(X, Y))

Chỉ số R-square =  0.6554631732894476


Chỉ số R-square =  0.6554631732894476 điều đó có nghĩa rằng các biến đầu vào đã giải thích được ~ 65.546% sự biến động của biến mục tiêu

Tính toán chỉ số MSE

Chúng ta có thể dự đoán đầu ra, tức là "yhat" bằng phương thức `.predict(X)`, trong đó X là biến đầu vào:

In [22]:
Yhat=lm.predict(X)
print('The output of the first four predicted value is: ', Yhat[0:5])

The output of the first four predicted value is:  [[14515.26092092]
 [14515.26092092]
 [21919.43979421]
 [12965.54906372]
 [15204.02174634]]


Bây giờ hãy import <b>mean_squared_error</b> từ module <b>metrics</b>


In [23]:
from sklearn.metrics import mean_squared_error

Chúng ta hãy so sánh kết quả dự đoán với kết quả thực tế

In [24]:
mse = mean_squared_error(Y, Yhat)
print('MSE giữa kết quả thực tế và kết quả dự đoán là: ', mse)

MSE giữa kết quả thực tế và kết quả dự đoán là:  21651264.45659664


<h3>Mô hình 2: Mô hình Multiple Linear Regression</h3>


Tính toán chỉ số R^2

In [25]:
# fit the model 
lm1 = LinearRegression()
lm1.fit(Z, Y)
# Find the R^2
print('Chỉ số R-square = ', lm1.score(Z, Y))

Chỉ số R-square =  0.811811561534475


Chỉ số R-square =  0.8118115615344751 điều đó có nghĩa rằng các biến đầu vào đã giải thích được ~ 81.181% sự biến động của biến mục tiêu

Tính toán chỉ số MSE

In [26]:
Y_predict_multifit = lm1.predict(Z)

Chúng ta hãy so sánh kết quả dự đoán với kết quả thực tế

In [27]:
print('MSE giữa kết quả thực tế và kết quả dự đoán là:  ', \
      mean_squared_error(Y, Y_predict_multifit))

MSE giữa kết quả thực tế và kết quả dự đoán là:   11826072.956532085


<h3>Mô hình 3: Mô hình Polynomial Fit (PMLR)</h3>


Tính toán chỉ số R^2

In [28]:
print('Chỉ số R-square = ', pipe.score(Z, Y))

Chỉ số R-square =  0.8531061720624729


Chỉ số R-square =  0.8531061720624729 điều đó có nghĩa rằng các biến đầu vào đã giải thích được ~ 85.31% sự biến động của biến mục tiêu

<h3>MSE</h3>


Tính toán chỉ số MSE

In [29]:
Y_poly_fit = pipe.predict(Z)
print('MSE giữa kết quả thực tế và kết quả dự đoán là: ', 
      mean_squared_error(Y, Y_poly_fit))

MSE giữa kết quả thực tế và kết quả dự đoán là:  9231051.281461732


<h3>Ra quyết định: Xác định mô hình nào là phù hợp nhất</h3>


<p> Bây giờ chúng ta đã xây dựng được 3 mô hình để dự đoán giá xe, tính toán được các chỉ số R-squared và MSE để đo lường mức độ phù hợp của mô hình, làm cách nào để xác định mô hình nào là phù hợp nhất?
    
<ul>
    <li><i>Một chỉ số R-squared tốt là như thế nào?</i></li>
</ul>
</p>

<p>Khi so sánh giữa các mô hình, <b>Mô hình nào có chỉ số R-squared càng lớn, mô hình đó càng fit với bộ dữ liệu dùng để huấn luyện.</b>.
<ul>
    <li><i>Một chỉ số MSE tốt là như thế nào?</i></li>
</ul>
</p>

<p>Khi so sánh giữa các mô hình, <b>Mô hình có chỉ số MSE bé nhất, mô hình đó hoạt động tốt nhất với bộ dữ liệu dùng để huấn luyện.</b></p>

<h4>Chúng ta hãy xem xét các giá trị cho các mô hình đã được huấn luyện:</h4>

<p>Mô hình Linear Regression đơn biến: Sử dụng biến Horsepower làm biến dự đoán giá.
<ul>
    <li>R-squared: 0.65546</li>
    <li>MSE: 2.16 x 10^7</li>
</ul>
</p>
    
<p>Mô hình Linear Regression đơn biến: sử dụng biến Horsepower, Curb-weight, Engine-size, và Highway-L/100km làm biến dự đoán giá.
<ul>
    <li>R-squared: 0.81181</li>
    <li>MSE: 1.2 x 10^7</li>
</ul>
</p>
    
<p>Mô hình Polynomial Fit: sử dụng biến Horsepower, Curb-weight, Engine-size, và Highway-L/100km làm biến dự đoán giá.
<ul>
    <li>R-squared: 0.85310</li>
    <li>MSE: 0.9 x 10^7</li>
</ul>
</p>

<h2>Conclusion:</h2>


So sánh kết quả của mô hình, chúng ta có thể kết luận rằng <b>Mô hình PMLR là mô hình tốt nhất</b> để dự đoán giá xe ô tô từ tập dữ liệu hiện có. 